src/discovery/discovery.utils.ts

```diff
@@ -0,0 +1,315 @@
+        [
+          ownerEOA,
+          config.recoverer,
+          config.logic,
+          config.logicParamsHash ?? SHA3_NULL_S,
+          walletIndex,
+        ]
+      )
+    : solidityKeccak256(
+        ['address', 'address', 'uint256'],
+        [ownerEOA, config.recoverer, walletIndex]
+      );
+
+  const data: string = solidityKeccak256(
+    ['bytes1', 'address', 'bytes32', 'bytes32'],
+    ['0xff', config.factory, salt, bytecodeHash]
+  );
+
+  const address = `0x${data.slice(26, data.length)}`;
+
+  return getAddress(address);
+};
+
+const checkHasActivity = async (
+  accountAddr: string,
+  config: Config,
+  provider: providers.Provider
+): Promise<boolean> => {
+  const hasBalance = (await provider.getBalance(accountAddr, 'latest')).gt(0);
+
+  if (hasBalance) {
+    return true;
+  }
+
+  const hasTx = (await provider.getTransactionCount(accountAddr, 'latest')) > 0;
+
+  if (hasTx) {
+    return true;
+  }
+
+  const hasERC20Balance = Boolean(
+    config.tokens.find(async (tokenAddr: string) => {
+      const erc20Contract = ERC20__factory.connect(tokenAddr, provider);
+      const balance = await erc20Contract.balanceOf(accountAddr);
+
+      return balance.gt(0);
+    })
+  );
+
+  if (hasERC20Balance) {
+    return true;
+  }
+
+  return false;
+};
+
+type Setup = {
+  config: Config;
+  rootNode?: HDNode;
+  bytecodeHash: string;
+  smartWalletFactory: SmartWalletFactory;
+  provider: providers.BaseProvider;
+};
+
+async function setupDiscovery(
+  configOverride: Config,
+  mnemonic?: string,
+  password?: string
+): Promise<Setup> {
+  const config = {
+    ...DEFAULT_DISCOVERY_CONFIG,
+    ...configOverride,
+  };
+
+  const rootNode = mnemonic
+    ? HDNode.fromMnemonic(
+        mnemonic,
+        password,
+        wordlists[config.mnemonicLanguage]
+      )
+    : undefined;
+  const provider = getDefaultProvider(config.network);
+
+  const smartWalletFactory = SmartWalletFactory__factory.connect(
+    config.factory,
+    provider
+  );
+
+  const creationByteCode = await smartWalletFactory.getCreationBytecode();
+  const bytecodeHash = keccak256(creationByteCode);
+  const logic = config.logic ?? constants.AddressZero;
+  const logicParamsHash = config.logicParamsHash ?? SHA3_NULL_S;
+
+  return {
+    config: {
+      ...config,
+      logic,
+      logicParamsHash,
+    },
+    rootNode,
+    bytecodeHash,
+    smartWalletFactory,
+    provider,
+  };
+}
+
+async function findHardenedNodeActivity(
+  { config, rootNode, provider, bytecodeHash }: Required<Setup>,
+  hardenedAccIdx: number
+): Promise<Account[]> {
+  let accounts: Account[] = [];
+  let inactivityCountdown = config.eoaGap;
+  let eoaIndex = 0;
+
+  while (inactivityCountdown > 0) {
+    const { address: eoaAccount } = rootNode.derivePath(
+      constructDerivationPath(eoaIndex, hardenedAccIdx)
+    );
+    const eoaActivityFound = await checkHasActivity(
+      eoaAccount,
+      config,
+      provider
+    );
+
+    if (eoaActivityFound) {
```

Comment:
why are you validating eoaActivityFound here?

Reply:
please expand your question. as it is I can only answer: why not?

Reply:
In line 243 you are validating the eoaActivityFound flag and restoring the gap. I don't think that the flag would change inside `findSWActivity`, therefore why is necessary to do this validating twice?

Reply:
ah well spotted tx

Reply:
done